### PR TITLE
Default values

### LINKED ISSUE #24: Support Default Values
- [ ] Hook `SqlTable::Insert` to widen the projected
- [ ] Hook `SqlTable::Select` and `SqlTable::Scan` to insert default values on old tables when appropriate (i.e. column added with default value)
- [ ] Hook `SqlTable::UpdateSchema` to tag old versions with default values when appropriate (i.e. column added with a default value)
- [ ] Modify `Schema::Column` to pass down default values (possibly as a fixed-length byte array and attribute width?  This would keep us from having to tracking and reasoning about underlying types)

### PR DESCRIPTION
**Pull request by saikiriti93**
Default values for MV tables. Fixes #24 

PS: My C++ is pretty bad! So please feel free to suggest improvements.

### CODE REVIEW

File: src/include/catalog/schema.h [hunk 1]
```diff
@@ -34,22 +34,48 @@ class Schema {
...
+    }
+
+    /**
+     * Instantiates a Column object, primary to be used for building a Schema object
+     * NOTE: Overloaded to avoid changing all Column initializations in the code
+     * @param name column name
+     * @param type SQL type for this column
+     * @param nullable true if the column is nullable, false otherwise
+     * @param oid internal unique identifier for this column
+     * @param default_value default value for this column
```
Comment (jrolli) [inline src/include/catalog/schema.h]: Where does the data to this pointer live?  Should this be a copy we are taking ownership of (avoids the memcpy below) or should this be immutable?

File: src/include/catalog/schema.h [hunk 2]
```diff
@@ -34,22 +34,48 @@ class Schema {
...
+     */
+    Column(std::string name, const type::TypeId type, const bool nullable, const col_oid_t oid, byte *default_value)
```
Comment (jrolli) [inline src/include/catalog/schema.h]: Can we overload the definitions by providing `nullptr` as the default value for `default_value` and using an if statement to do the default_value assignment?  This would save the work of checking for `nullptr` below and give a clean semantic that the "default value" is normally `null`.

File: src/include/catalog/schema.h [hunk 3]
```diff
@@ -34,22 +34,48 @@ class Schema {
...
           oid_(oid) {
-      if (attr_size_ == VARLEN_COLUMN) {
-        // this is a varlen attribute
-        // attr_size_ is actual size + high bit via GetTypeSize
-        inlined_ = false;
+      // ASSUMPTION: The default_value passed in is of size attr_size_
+      // Copy the passed in default value
+      if (default_value != nullptr) {
+        default_ = new byte[attr_size_];
+        std::memcpy(default_, default_value, attr_size_);
```
Comment (jrolli) [inline src/include/catalog/schema.h]: We may need to reach out and reason more about how VARLEN actually works.  I think there is a reference counting mechanism that we may need to interact with but I'm not sure.

File: src/include/catalog/schema.h [hunk 4]
```diff
@@ -34,22 +34,48 @@ class Schema {
...
+      } else {
+        default_ = default_value;
```
Comment (jrolli) [inline src/include/catalog/schema.h]: Write as `default_ = nullptr;`. It's easier to read and equivalent since we know what `default_value` is.  Also, I'd personally move this up to line 64 and get rid of the `else` clause entirely, but that's more of a style preference.

File: src/include/storage/sql_table.h [hunk 5]
```diff
@@ -32,7 +32,8 @@ class SqlTable {
     BlockLayout layout;
     ColumnMap column_map;
     InverseColumnMap inverse_column_map;
-    // TODO(John): Add 'default_value_map' (dynamic) for col_oid->default_val
+    // TODO(Sai): Do we really need a defaultValueMap for every version? One for SqlTable should work?
```
Comment (jrolli) [inline src/include/storage/sql_table.h]: I think you may be correct, but I would like to whiteboard it out to make sure.   One thing that would become more difficult, though is cleaning up old default values that we don't need to hang onto anymore (i.e. compactor has come through and we no longer have datatables created before the column with default was added).

File: test/storage/sql_table_test.cpp [hunk 6]
```diff
@@ -526,9 +544,10 @@ TEST_F(SqlTableTests, ScanTest) {
...
   table.version_ = storage::layout_version_t(1);
-  table.AddColumn(txn, "new_col", type::TypeId::INTEGER, true, catalog::col_oid_t(2));
+  table.AddColumn(txn, "new_col", type::TypeId::INTEGER, true, catalog::col_oid_t(2),
+                  table.intToByteArray(new_col_default_value));
 
-  // insert (300, 10002, null)
+  // insert (300, 10002, 1729)
   table.StartInsertRow();
   table.SetIntColInRow(catalog::col_oid_t(0), 300);
   table.SetIntColInRow(catalog::col_oid_t(1), datname_map[300]);
```
Comment (jrolli) [inline test/storage/sql_table_test.cpp:663]: Column with default value added before this insertion started.  I'm pretty sure the expectation is that the query plan would handle setting the default value, not us.

File: test/storage/sql_table_test.cpp [hunk 7]
```diff
@@ -570,10 +589,10 @@ TEST_F(SqlTableTests, ScanTest) {
   uint32_t datname = *reinterpret_cast<uint32_t *>(value);
   EXPECT_EQ(datname, datname_map[id]);
   value = row1.AccessWithNullCheck(pc_pair.second.at(catalog::col_oid_t(2)));
+  uint32_t new_col = *reinterpret_cast<uint32_t *>(value);
```
Comment (jrolli) [inline test/storage/sql_table_test.cpp:703]: Based off default value semantics, I would expect the following mapping from `ID` to `new_col`:
```
  100 -> new_col_default_value
  200 -> new_col_default_value
  300 -> nullptr
  400 -> new_col_map[id]
```

Specifically, the assumption would be that the `NULL` in 300 is because the query explicitly overrode the default value with `NULL` and the query planner somehow optimized away the insert value.

File: test/storage/sql_table_test.cpp [hunk 8]
```diff
@@ -570,10 +589,10 @@ TEST_F(SqlTableTests, ScanTest) {
...
   if (id != 400) {
-    EXPECT_EQ(value, nullptr);
+    EXPECT_EQ(new_col, new_col_default_value);
```
Comment (jrolli) [inline test/storage/sql_table_test.cpp:705]: See above

File: test/storage/sql_table_test.cpp [hunk 9]
```diff
@@ -589,10 +608,10 @@ TEST_F(SqlTableTests, ScanTest) {
   datname = *reinterpret_cast<uint32_t *>(value);
   EXPECT_EQ(datname, datname_map[id]);
   value = row2.AccessWithNullCheck(pc_pair.second.at(catalog::col_oid_t(2)));
+  new_col = *reinterpret_cast<uint32_t *>(value);
   if (id != 400) {
-    EXPECT_EQ(value, nullptr);
+    EXPECT_EQ(new_col, new_col_default_value);
```
Comment (jrolli) [inline test/storage/sql_table_test.cpp:724]: See above

File: test/storage/sql_table_test.cpp [hunk 10]
```diff
@@ -613,10 +632,10 @@ TEST_F(SqlTableTests, ScanTest) {
   datname = *reinterpret_cast<uint32_t *>(value);
   EXPECT_EQ(datname, datname_map[id]);
   value = row3.AccessWithNullCheck(pc_pair.second.at(catalog::col_oid_t(2)));
+  new_col = *reinterpret_cast<uint32_t *>(value);
   if (id != 400) {
-    EXPECT_EQ(value, nullptr);
+    EXPECT_EQ(new_col, new_col_default_value);
```
Comment (jrolli) [inline test/storage/sql_table_test.cpp:748]: See above

File: test/storage/sql_table_test.cpp [hunk 11]
```diff
@@ -632,10 +651,10 @@ TEST_F(SqlTableTests, ScanTest) {
   datname = *reinterpret_cast<uint32_t *>(value);
   EXPECT_EQ(datname, datname_map[id]);
   value = row4.AccessWithNullCheck(pc_pair.second.at(catalog::col_oid_t(2)));
+  new_col = *reinterpret_cast<uint32_t *>(value);
   if (id != 400) {
-    EXPECT_EQ(value, nullptr);
+    EXPECT_EQ(new_col, new_col_default_value);
```
Comment (jrolli) [inline test/storage/sql_table_test.cpp:767]: See above

File: test/storage/sql_table_test.cpp [hunk 12]
```diff
@@ -489,6 +506,7 @@ TEST_F(SqlTableTests, UpdateTest) {
 TEST_F(SqlTableTests, ScanTest) {
   std::map<uint32_t, uint32_t> datname_map;
   std::map<uint32_t, uint32_t> new_col_map;
+  uint32_t new_col_default_value = 1729;
```
Comment (jrolli) [inline test/storage/sql_table_test.cpp:619]: I think we should draw default values out into a separate test suite.  There are several edge cases and sequencing concerns that I think we need to explicitly test.  Specifically, things like: (1) column added without a default value has one in a later schema change (we should ignore this since it is not retroactive), (2) column added with a default value does not have one in a later schema change (we need to maintain the old default and continue to apply it to the tables that don't have the column), (3) VARLEN default values for an added column that cannot be inlined (i.e. more than 12 characters long) [has possible reference counting/correctness concerns], and possibly others.

File: src/storage/sql_table.cpp [hunk 13]
```diff
@@ -108,6 +108,16 @@ void SqlTable::UpdateSchema(const catalog::Schema &schema) {
     }
   }
 
+  // Populate the default value map
```
Comment (jrolli) [inline src/storage/sql_table.cpp:78]: I think we actually need to modify every *other* datatable's default value map.  Specifically, if this schema (1) has a column OID the previous schema version does not and (2) that column comes with a default value, then (3) we need to apply that default value to all previous schema versions to ensure "new" reads will see the retroactive value.

File: src/storage/sql_table.cpp [hunk 14]
```diff
@@ -145,7 +156,20 @@ bool SqlTable::Select(transaction::TransactionContext *const txn, const TupleSlo
   bool result = old_dt_version.data_table->Select(txn, slot, out_buffer);
   std::memcpy(out_buffer->ColumnIds(), original_column_ids, sizeof(col_id_t) * out_buffer->NumColumns());
 
-  // TODO(Yashwanth): handle default values
+  // Get the DefaultValueMap visible to this transaction
+  DefaultValueMap default_val_map = curr_dt_version.default_value_map;
+
+  // Populate default values into the empty columns of the ProjectedRow (if any)
```
Comment (jrolli) [inline src/storage/sql_table.cpp]: See the comment in `UpdateSchema`.  This is more complicated then just checking for `NULL` since default values are only retroactive if they are applied when a column is first added.  Additionally, you can override a default value by explicitly setting the value to `NULL` (if constraints allow) in the insert statement.

File: src/storage/sql_table.cpp [hunk 15]
```diff
@@ -287,6 +311,36 @@ void SqlTable::Scan(transaction::TransactionContext *const txn, SqlTable::SlotIt
   uint32_t filled = out_buffer->NumTuples();
   std::memcpy(out_buffer->ColumnIds(), original_column_ids, sizeof(col_id_t) * out_buffer->NumColumns());
   out_buffer->SetNumTuples(filled);
+
+  // Populate the default values
```
Comment (jrolli) [inline src/storage/sql_table.cpp]: See comment in `Select`.

File: src/storage/sql_table.cpp (hunk 13 above)
Comment (jrolli) [reply src/storage/sql_table.cpp:78]: We should probably pass in the transaction context and transaction version to this function make this check easier and have a more consistent API across `SqlTable`.

**Review by jrolli — CHANGES_REQUESTED** [review]
The code to apply default values looks clean, but I think the code applies default values in cases where it shouldn't.  A test suite just for default values would help keep us on track.

File: src/include/catalog/schema.h (hunk 1 above)
Comment (saikiriti93) [reply src/include/catalog/schema.h]: I was thinking this can be immutable so that the layers above it doesn't need to worry about this.

File: src/include/catalog/schema.h (hunk 2 above)
Comment (saikiriti93) [reply src/include/catalog/schema.h]: Sorry I didn't understand it. You are implying that the default value should be passed on to the constructor always? If yes, why would we need the `if` condition?

File: test/storage/sql_table_test.cpp (hunk 7 above)
Comment (saikiriti93) [reply test/storage/sql_table_test.cpp:703]: Forcing the value to be `NULL` despite the presence of a default value would be problematic right? How do we know now about which columns were actually `NULL` and which columns were set to `NULL` by the `select`/`scan` functions? For `select` I can see passing in a vector or something which tells the columns it had filled in. But for `scan` we need to maintain the columns filled in for each of the tuples to do this.

File: test/storage/sql_table_test.cpp (hunk 12 above)
Comment (saikiriti93) [reply test/storage/sql_table_test.cpp:619]: Yeah, you are right. I'll make more comprehensive test cases for the default values. I have a few questions about the points you made though:
1) How is this change of default value going to be implemented? Should I be exposing a `setDefaultValue` method? In that case, I have to find some other flag to indicate whether the default value in the `Column` is assigned at the time of creation or later.
2) Again, should I expose an API like `clearDefaultValue`? But this shouldn't actually clear the default values from the `DefaultValueMap` in the `DataTableVersion`.
3) I'll check the code about how VARLEN reference counting is handled. It'll be a very bad idea to copy the VARLEN content as I'm doing it now.

File: src/storage/sql_table.cpp (hunk 13 above)
Comment (saikiriti93) [reply src/storage/sql_table.cpp:78]: > Specifically, if this schema (1) has a column OID the previous schema version does not

I'm using the `DefaultValueMap` of the version as seen by the transaction while filling in default values. So, the previous version's default values don't matter, right? Also, if a new column OID comes in which the previous schema versions don't contain, they shouldn't be made aware of the new column OID.

File: src/storage/sql_table.cpp (hunk 14 above)
Comment (saikiriti93) [reply src/storage/sql_table.cpp]: The defaultValueMap that I'm constructing is assumed to contain only the values that should be applied retroactively. I've assumed that the catalog maintains the default values that are applied after the column creation at another place.

If the default value in the `Column` class is the only copy, then I need to maintain a flag indicating whether this default value was assigned during creation or later.

File: src/include/storage/storage_defs.h [hunk 16]
```diff
@@ -175,8 +175,12 @@ using ColumnMap = std::unordered_map<catalog::col_oid_t, col_id_t>;
 /**
  * Used by execution and storage layers to map between col_oids and offsets within a ProjectedRow
  */
-using InverseColumnMap = std::unordered_map<col_id_t, catalog::col_oid_t>;
 using ProjectionMap = std::unordered_map<catalog::col_oid_t, uint16_t>;
+using InverseColumnMap = std::unordered_map<col_id_t, catalog::col_oid_t>;
+/**
+ * Used by SqlTable to map between col_oids in Schema and their {default_value, attribute_size}
+ */
+using DefaultValueMap = std::unordered_map<catalog::col_oid_t, std::pair<byte *, uint8_t>>;
```
Comment (jrolli) [inline src/include/storage/storage_defs.h]: We should probably use a concurrent unordered map here.  Even though adding is single-threaded there are still potential lookups happening concurrently with that operation.

File: test/storage/sql_table_test.cpp [hunk 17]
```diff
@@ -791,4 +832,172 @@ TEST_F(SqlTableTests, MultipleColumnWidths) {
   txn_manager_.Commit(txn, TestCallbacks::EmptyCallback, nullptr);
   delete txn;
 }
+
+// NOLINTNEXTLINE
+TEST_F(SqlTableTests, BasicDefaultValuesTest) {
+  // TODO(Sai): Merge the sql_table_test and sql_table_concurrent_test frameworks to avoid repetition for adding in
+  // columns. Have 3 types of columns, 1 with default value at creation, one without and one with default value added
```
Comment (jrolli) [inline test/storage/sql_table_test.cpp]: Does this comment need to be updated?  I think you changed the test based off of our conversation (SqlTable wouldn't see a default value change that occurred after column creation), but this says you're testing it.

Comment (saikiriti93) [reply test/storage/sql_table_test.cpp]: Yes, it was outdated.

File: src/include/storage/storage_defs.h (hunk 16 above)
Comment (saikiriti93) [reply src/include/storage/storage_defs.h]: You are right. I've changed it to common::ConcurrentMap
